core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java

```diff
@@ -0,0 +1,224 @@
+
+    /**
+     * Parameter passed to {@link #setUp} to allow an implementation to specify its behavior after the initial setup.
+     */
+    public static final class Context {
+        private Disposer disposer;
+        private final Map<String,String> env = new HashMap<String,String>();
+        public @Nonnull Map<String,String> getEnv() {
+            return env;
+        }
+        /**
+         * Specify an environment variable override to apply to processes launched within the block.
+         * If unspecified, environment variables will be inherited unmodified.
+         * @param key handles the special {@code PATH+SOMETHING} syntax as in {@link EnvVars#override}
+         */
+        public void env(String key, String value) {
+            if (env.containsKey(key)) {
+                throw new IllegalStateException("just one binding for " + key);
+            }
+            env.put(key, value);
+        }
+        public @CheckForNull Disposer getDisposer() {
+            return disposer;
+        }
+        /**
+         * Specify an action to take when the block ends.
+         * If not specified, nothing special happens.
+         */
+        public void setDisposer(@Nonnull Disposer disposer) {
+            if (this.disposer != null) {
+                throw new IllegalStateException("just one disposer");
+            }
+            this.disposer = disposer;
+        }
+    }
+
+    /**
+     * An optional callback to run at the end of the wrapped block.
+     * Must be safely serializable, so it receives runtime context comparable to that of the original setup.
+     */
+    public static abstract class Disposer implements Serializable {
+        /**
+         * Attempt to clean up anything that was done in the initial setup.
+         * @param build a build being run
+         * @param workspace a workspace of the build
+         * @param launcher a way to start commands
+         * @param listener a way to report progress
+         * @throws IOException if something fails; {@link AbortException} for user errors
+         * @throws InterruptedException if tear down is interrupted
+         */
+        public abstract void tearDown(Run<?,?> build, FilePath workspace, Launcher launcher, TaskListener listener) throws IOException, InterruptedException;
+    }
+
+    /**
+     * By default, when run as part of an {@link AbstractBuild}, will run late, in the {@link #setUp(AbstractBuild, Launcher, BuildListener)} phase.
+     * May be overridden to return true, in which case this will run earlier, in the {@link #preCheckout} phase.
+     * Ignored when not run as part of an {@link AbstractBuild}.
+     */
+    protected boolean runPreCheckout() {
+        return false;
+    }
+
+    @Override public final Environment setUp(AbstractBuild build, final Launcher launcher, BuildListener listener) throws IOException, InterruptedException {
+        if (runPreCheckout()) {
+            return new Environment() {};
```

Comment:
The reference is harmless however, because in this context the `Environment` is not serialized, and the `BuildWrapper` will last at least as long (the `buildEnvironments` are cleared in `AbstractBuildExecution.cleanUp`).